### ISSUE TITLE
How to operate beringei in production

### ISSUE THREAD

**Issue by huydx** [facebookincubator/beringei#61]
I found that the document is lacking of guide to know what is the good way to operate beringei at production, where we need multi nodes and a good way to sharding.
I may miss a good document elsewhere, so if you have a good one, please tell me.

Thank you.

**Comment by linuxfood** [comment]
Heya @huydx, this is definitely an area that could use some more documentation.
The tl;dr is that right now it's a bit of an "exercise left to the reader." What we use in production at FB is not quite what we've open sourced here, unfortunately. We have some internal frameworks we're not able to open source right now that handle some aspects of this problem for us.

Still, I can give you some guidance and make a note for us to document this more clearly.

**First**, you're going to need some sort of shared storage that exposes some sort of POSIX filesystem API. We currently use NFS for this, but any number of other POSIX-y distributed filesystems should work as well. In particular though, you can't use HDFS for this, as it can't be mounted (to the best of my knowledge) as a filesystem. You need this to be mounted on all hosts that are running Beringei. They use this for storage hand-off when there is shard movement.

**Second**, you need some way of maintaining a shard map (service configuration). This is the JSON configuration file that the README shows you how to generate. It's up to you to decide how best to balance the shards across available machines, and to figure out how to synchronize the file across machines. If you're not afraid to write a little bit of code, you could make a subclass of `beringei/client/BeringeiConfigurationAdapterIf.h` that could load configuration from Zookeeper, Etcd, or Consul. Though, disclaimer: We don't have any direct experience using these technologies in concert with Beringei.

Feel free to comment here if you have any more questions!

**Comment by linuxfood** [comment]
Hi @huydx 

Sorry for the extreme latency in my reply. We somehow missed the notification that you had commented with more questions!

> How do you handle multi-tenant application (are you using single big beringei cluster, or one beringei per app?)

Beringei was designed from the beginning to be able to handle multi-tenancy. So, while we have several different deployments around the world, they all hold a complete replica of the monitoring data. This reduces the application logic necessary to find a piece of data, and makes the operations of running beringei easier overall.

> In case you're handling multi-tenant application, what is your sharding key, and how to make it efficient at read operation.

We just apply a hash function to the key, with a modulo of our shard count to compute the shard id. As far as efficiency goes, I don't think we've ever had a problem with reads. We have some systems that actually consume entire shards at a time to compute complex derived data for whole time-series at a time. Beringei is able to service these requests just fine along with anything else. Do you have a specific use case you're trying to ensure it can handle?

> Are you handling multi dimensional data (in another db system: it just tag, or label). If yes, how you handle that with beringei which supports single key name only AFAIK.

Yeah, we handle multi-dimensional data. In our production deployment the key in Beringei is the concatenation of a few different pieces of information that allows us to uniquely identify a time-series. We don't use tags or labels specifically, but, as long as you can ensure stable key names, and enumerate them somehow, then you can add whatever kind of abstraction you want over Beringei's keys.
